[iOS] Introduce MediaDeviceRoute

aestes opened the pull request:
#### cbd19cda5899cc9b5b8f220e398b307e7bb02c8b
<pre>
[iOS] Introduce MediaDeviceRoute
<a href="https://bugs.webkit.org/show_bug.cgi?id=303214">https://bugs.webkit.org/show_bug.cgi?id=303214</a>
<a href="https://rdar.apple.com/165498830">rdar://165498830</a>

Reviewed by Jean-Yves Avenard.

Introduced MediaDeviceRoute. In a follow-on change it will be used by
MediaPlayerPrivateWirelessPlayback for playing media to AirPlay routes. MediaDeviceRoute&apos;s job is
to wrap a WebMediaDevicePlatformRoute, observing its properties and dispatching
MediaDeviceRouteClient callbacks when values change. It also exposes a C++ interface for getting
and setting route properties.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/ios/MediaDeviceRoute.h: Added.
(WebCore::MediaDeviceRoute::client const):
(WebCore::MediaDeviceRoute::setClient):
* Source/WebCore/platform/audio/ios/MediaDeviceRoute.mm: Added.
(-[WebMediaDeviceRoute initWithRoute:platformRoute:]):
(-[WebMediaDeviceRoute platformRoute]):
(-[WebMediaDeviceRoute observeValueForKeyPath:ofObject:change:context:]):
(-[WebMediaDeviceRoute dealloc]):
(WebCore::convert):
(WebCore::MediaDeviceRoute::create):
(WebCore::MediaDeviceRoute::MediaDeviceRoute):
* Source/WebCore/platform/cocoa/KeyEventCocoa.mm:
(WebCore::keyForCharCode):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.h:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDTMFSenderBackend.cpp:

Canonical link: <a href="https://commits.webkit.org/303660@main">https://commits.webkit.org/303660@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5bcfbaa53323270bb2bccf7a66ffc5cecff3d0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133022 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44120 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140558 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85055 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4e0fffb7-a082-4654-9f78-2eb156676e85) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134892 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5386 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101725 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69040 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135968 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4198 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119197 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82524 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3bb61972-ded8-4d1b-9334-f0201d571295) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4081 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1686 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83793 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113140 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37313 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143211 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5192 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37895 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110103 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5274 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110285 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27985 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3987 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115458 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58779 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5246 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33808 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5087 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68698 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5336 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5204 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/audio/ios/MediaDeviceRoute.h:
```diff
@@ -0,0 +1,162 @@
+    WTF_MAKE_TZONE_ALLOCATED(MediaDeviceRoute);
+public:
+    static Ref<MediaDeviceRoute> create(WebMediaDevicePlatformRoute *);
+
+    ~MediaDeviceRoute();
```
```suggestion
    virtual ~MediaDeviceRoute();
```

jyavenard replied:
or make class final.

jyavenard commented on Source/WebCore/platform/audio/ios/MediaDeviceRoute.mm:
```diff
@@ -0,0 +1,359 @@
+    { \
+        return convert([[m_route platformRoute] KeyPath]); \
+    } \
+\
```
needed this last `\` ?

aestes replied:
No, I can remove these.

jyavenard commented on Source/WebCore/platform/audio/ios/MediaDeviceRoute.mm:
```diff
@@ -0,0 +1,359 @@
+    { \
+        [[m_route platformRoute] set##SetterSuffix:convert(WTFMove(KeyPath))]; \
+    } \
+\
```
ditto

jyavenard commented on Source/WebCore/platform/audio/ios/MediaDeviceRoute.mm:
```diff
@@ -0,0 +1,359 @@
+    }
+
+    FOR_EACH_KEY_PATH(OBSERVE_VALUE)
+    ASSERT_NOT_REACHED();
```
that ASSERT_NOT_REACHED() seems wrong here ; aren't we always going to reach here?

aestes replied:
Every `OBSERVE_VALUE` block returns. If we made it here, that means we observed a value change for an unexpected property, which would be programmer error.

jyavenard commented on Source/WebCore/platform/audio/ios/MediaDeviceRoute.mm:
```diff
@@ -0,0 +1,359 @@
+
+static double convert(double value)
+{
+    return value;
```
I'm a bit confused on what those two converts methods are for...

Plus a bool can be converted to a double, doesn't this generate warnings ?

aestes replied:
Exact matches are preferred to matches that require promotion, so there's no ambiguity here. It's a little silly, but these are here because `DEFINE_SETTER` unconditionally calls `convert` on every argument.

aestes replied:
(but maybe it would be better to use a function template with specializations for the non-trivial cases; I can look into that as a follow-up)

jyavenard reviewed (approved):
r=me with nits

aestes committed:
[iOS] Introduce MediaDeviceRoute
https://bugs.webkit.org/show_bug.cgi?id=303214
rdar://165498830

Reviewed by Jean-Yves Avenard.

Introduced MediaDeviceRoute. In a follow-on change it will be used by
MediaPlayerPrivateWirelessPlayback for playing media to AirPlay routes. MediaDeviceRoute's job is
to wrap a WebMediaDevicePlatformRoute, observing its properties and dispatching
MediaDeviceRouteClient callbacks when values change. It also exposes a C++ interface for getting
and setting route properties.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/audio/ios/MediaDeviceRoute.h: Added.
(WebCore::MediaDeviceRoute::client const):
(WebCore::MediaDeviceRoute::setClient):
* Source/WebCore/platform/audio/ios/MediaDeviceRoute.mm: Added.
(-[WebMediaDeviceRoute initWithRoute:platformRoute:]):
(-[WebMediaDeviceRoute platformRoute]):
(-[WebMediaDeviceRoute observeValueForKeyPath:ofObject:change:context:]):
(-[WebMediaDeviceRoute dealloc]):
(WebCore::convert):
(WebCore::MediaDeviceRoute::create):
(WebCore::MediaDeviceRoute::MediaDeviceRoute):
* Source/WebCore/platform/cocoa/KeyEventCocoa.mm:
(WebCore::keyForCharCode):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.h:
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDTMFSenderBackend.cpp:

Canonical link: https://commits.webkit.org/303660@main
